### PR TITLE
Fix Snyk scanning and upload results

### PR DESCRIPTION
**Pull request by julienp**
Fix scanning for all images by correctly setting the platform. Scan results are upload to GitHub code scanning.

### CODE REVIEW

**Comment by julienp** [comment]
This currently flags a number of issues https://github.com/pulumi/pulumi-docker-containers/security/code-scanning?page=2&query=is%3Aopen+branch%3Ajulienp%2Fsnyk

We do get a fair amount of duplication across the images. Most of them are the recent `golang.org/x/crypto/ssh` issue, which should be fixed in the next pulumi release by virtue of building with the latest golang. There's a number of other things in there I will triage once this lands.

File: .github/workflows/snyk-scan.yml [hunk 1]
```diff
@@ -17,14 +17,46 @@ jobs:
...
+          SNYK_TOKEN: ${{ secrets.SNYK_TOKEN }}
+        with:
+          image: ${{ env.DOCKER_ORG }}/pulumi:${{ env.PULUMI_VERSION }}${{ matrix.suffix }}-amd64
+          args: --severity-threshold=high --file=docker/pulumi/Dockerfile
+      - name: Filter Sarif File
+        # GitHub Code Scanning does not allow more than 20 runs per file. We
+        # filter out empty runs to avoid this limit. Note that the file needs to
+        # include at least one run.
+        run: |
+          python -c 'import json; sarif = json.loads(open("snyk.sarif").read()); runs = [run for run in sarif["runs"] if len(run["results"]) > 0]; sarif["runs"] = runs if len(runs) > 0 else [sarif["runs"][0]]; open("out.sarif", "w").write(json.dumps(sarif, indent=2))'
```
Comment (lunaris) [inline .github/workflows/snyk-scan.yml:45]: Not _terrible_ but wondering if we want a `scripts/whatever.py` in lieu of this.

File: .github/workflows/snyk-scan.yml [hunk 2]
```diff
@@ -17,14 +17,46 @@ jobs:
...
+
+  provider-build-environment:
+    name: Provider Build Environment
+    runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@master
+      - name: Free Disk Space (Ubuntu)
+        uses: jlumbroso/free-disk-space@main
+        with:
+          tool-cache: false
```
Comment (lunaris) [inline .github/workflows/snyk-scan.yml:59]: What leads to tons of disk usage here? Is it the outputs of scanning?

File: .github/workflows/snyk-scan.yml [hunk 3]
```diff
@@ -17,14 +17,46 @@ jobs:
     name: All SDKs images
     runs-on: ubuntu-latest
     strategy:
+      fail-fast: false
```
Comment (lunaris) [inline .github/workflows/snyk-scan.yml:20]: I suspect there's a good reason for this -- worth documenting?

File: .github/workflows/snyk-scan.yml [hunk 4]
```diff
@@ -90,7 +144,18 @@ jobs:
...
         with:
           image: ${{ env.IMAGE_NAME }}
-          args: --severity-threshold=high --file=docker/${{ matrix.sdk }}/Dockerfile.debian
+          args: --severity-threshold=high --file=docker/${{ matrix.sdk }}/Dockerfile.debian --platform=linux/${{ matrix.arch }}
+      - name: Filter Sarif File
+        # GitHub Code Scanning does not allow more than 20 runs per file. We
+        # filter out empty runs to avoid this limit. Note that the file needs to
+        # include at least one run.
+        run: |
+          python -c 'import json; sarif = json.loads(open("snyk.sarif").read()); runs = [run for run in sarif["runs"] if len(run["results"]) > 0]; sarif["runs"] = runs if len(runs) > 0 else [sarif["runs"][0]]; open("out.sarif", "w").write(json.dumps(sarif, indent=2))'
```
Comment (lunaris) [inline .github/workflows/snyk-scan.yml:153]: With the duplication it does feel like a Python script in the repo could be valuable, if that's possible.

**Review by lunaris — APPROVED** [review]

File: .github/workflows/snyk-scan.yml (hunk 2 above)
Comment (julienp) [reply .github/workflows/snyk-scan.yml:59]: The image itself is fairly large, plus it download some snyk image. Not sure what precisely causes us to run out of space, but for kitchen-sink it failed with `ENOSPC` before :/

File: .github/workflows/snyk-scan.yml (hunk 1 above)
Comment (julienp) [reply .github/workflows/snyk-scan.yml]: Definitely, done

File: .github/workflows/snyk-scan.yml (hunk 3 above)
Comment (julienp) [reply .github/workflows/snyk-scan.yml:21]: Yeah we want to complete as many scans as possible, if one fails for some reason. I added a comment.

**Pull request merged**